mmdet/models/anchor_heads/ssd_head.py

```diff
@@ -21,77 +21,114 @@ def __init__(self,
-            indices = list(range(len(ratios)))
-            indices.insert(1, len(indices))
-            anchor_generator.base_anchors = torch.index_select(
-                anchor_generator.base_anchors, 0, torch.LongTensor(indices))
-            self.anchor_generators.append(anchor_generator)
+        if len(anchor_heights):
+            assert len(anchor_heights) == len(anchor_widths)
+            for k in range(len(anchor_strides)):
+                assert len(anchor_widths[i]) == len(anchor_heights[i])
+                stride = anchor_strides[k]
+                if isinstance(stride, tuple):
+                    ctr = ((stride[0] - 1) / 2., (stride[1] - 1) / 2.)
+                else:
+                    ctr = ((stride - 1) / 2., (stride - 1) / 2.)
+                anchor_generator = AnchorGenerator(
+                    0, [], [], widths=anchor_widths[k],
+                    heights=anchor_heights[k],
+                    scale_major=False, ctr=ctr)
+                self.anchor_generators.append(anchor_generator)
+        else:
+            min_ratio, max_ratio = basesize_ratio_range
+            min_ratio = int(min_ratio * 100)
+            max_ratio = int(max_ratio * 100)
+            step = int(np.floor(max_ratio - min_ratio) /
+                       (len(in_channels) - 2))
+            min_sizes = []
+            max_sizes = []
+            for r in range(int(min_ratio), int(max_ratio) + 1, step):
+                min_sizes.append(int(input_size * r / 100))
+                max_sizes.append(int(input_size * (r + step) / 100))
+            min_sizes.insert(0, int(input_size * basesize_ratio_range[0] / 2))
+            max_sizes.insert(0, int(input_size * basesize_ratio_range[0]))
+            for k in range(len(anchor_strides)):
+                base_size = min_sizes[k]
+                stride = anchor_strides[k]
+                if isinstance(stride, tuple):
+                    ctr = ((stride[0] - 1) / 2., (stride[1] - 1) / 2.)
+                else:
+                    ctr = ((stride - 1) / 2., (stride - 1) / 2.)
+                scales = [1., np.sqrt(max_sizes[k] / min_sizes[k])]
+                ratios = [1.]
+                for r in anchor_ratios[k]:
+                    ratios += [1 / r, r]  # 4 or 6 ratio
+                anchor_generator = AnchorGenerator(
+                    base_size, scales, ratios, scale_major=False, ctr=ctr)
+                indices = list(range(len(ratios)))
+                indices.insert(1, len(indices))
+                anchor_generator.base_anchors = torch.index_select(
+                    anchor_generator.base_anchors, 0,
+                    torch.LongTensor(indices))
+                self.anchor_generators.append(anchor_generator)
 
         self.target_means = target_means
         self.target_stds = target_stds
         self.use_sigmoid_cls = False
         self.cls_focal_loss = False
+        self.loss_balancing = loss_balancing
```

Comment:
Hi, can you provide a more detail ablation study about the loss balancing method from Kendal, etc? We have also tried the method on Faster R-CNN and other detectors but found unstable results (can't work end-to-end but can indeed learn a useful loss weight that can be used offline and bring improvements vs. baseline). If the method indeed helps in some cases, we hope the method can be implemented in loss evaluator as a general method.

Reply:
I've tested this method with SSD only. On WIDER dataset it helps to adapt tiny detector for big faces and generally prevents from overfitting on particular object size.
This scheme is not a plug-and-play method and requires tuning of optimization strategy. At least one need to increase learning rate to maintain the same amplitude of gradients.
Also this balancing method does each adjustment of weights based on previous minibatch. May be for huge detectors when the batch size is 2-4 this delay is critical and batches are too different on early stages, so such adjustment makes training unstable.